### PR TITLE
Support mlx-vlm 0.7.1 and mlx 0.32.2

### PR DESCRIPTION
**Pull request by Lyxot**
## Summary

Raises the supported MLX stack to mlx-vlm 0.7.1 and mlx 0.32.2, and fixes the two ways 0.7.1 breaks the MLX backend today. Both breaks are silent: one fails training outright, the other quietly stops accelerating.

## Qwen3.5 VLM training crashes on mlx-vlm 0.7.1

mlx-vlm keeps adding keywords to `gated_delta_update` — 0.6.9 added `lower_bound`, 0.7.1 added `cache`, `cache_index` and `state_steps` — and its call sites pass the new ones unconditionally, even when they are `None`. The patched wrappers that install the memory-efficient VJP had fixed signatures, so on 0.7.1 the call raised `TypeError` and Qwen3.5 VLM training could not start.

The wrappers now accept whatever arrives and splat it straight back when they delegate: a keyword the caller supplied is one the installed mlx-vlm accepts, which keeps a single wrapper correct from 0.4.x through 0.7.x without sniffing versions. Calls that own a cache, or that carry a keyword the training branch does not cover, delegate to upstream instead of silently training on a path the custom VJP does not model; an unrecognized keyword warns once and delegates rather than failing.

This also fixes a real staleness bug that predates 0.7.1: consumer modules that `from`-import `gated_delta_update` at import time kept the unpatched function, so patching the source module alone never reached them. The patch now sweeps `mlx_vlm.models.*` and rebinds by identity.

## MoE gate/up fusion silently stops applying on mlx-vlm 0.7.1

0.7.1 gave `SwitchGLU.__call__` a short-sequence decode path, taken when `not self.training and x.ndim == 3 and 1 < x.shape[1] <= DECODE_BLOCK_SIZE`, and widened the signature to `(x, indices, weights=None, shared=None, residual=None)`. mlx-lm 0.31.3 has neither, so the two switch layer packages no longer carry the same body and the single shared source fingerprint drifted. Output stayed correct because the guard falls back to the native path, but the packed gate/up projection stopped being used for every mlx-vlm MoE model.

Re-pinning the fingerprint alone is wrong, and is wrong in a way that only a bitwise check catches. Measured against the installed 0.7.1 across every quantization mode the fusion covers (affine 4/6/8, mxfp4, mxfp8, nvfp4) and sequence lengths 1, 2, 4, 8, 9, 16 and 64:

- the fused output is bitwise identical to upstream's own sort path at every length;
- upstream's new decode path is not bitwise equal to that same sort path;
- the two disagree only at length 8, where the fusion's `indices.size >= 64` sort threshold overlaps the decode window.

So the reimplementation is faithful and only has to stop answering for calls 0.7.1 routes elsewhere. `SwitchGLU.__call__` is now fingerprinted per package, with mlx-vlm accepting the pre-0.7.1 body as well so the whole supported range keeps fusing, and the fused call defers to the native one inside the decode window and whenever routing weights, a shared expert output or residuals arrive. The window is read from the package's own `DECODE_BLOCK_SIZE`, which is absent on mlx-lm and before 0.7.1 and reads as 0 there, leaving the guard inert.

## Dependency ceilings

`mlx-vlm>=0.4.4,<0.7.0` becomes `<=0.7.1` and `mlx==0.32.1` becomes `0.32.2`, in both the platform-marked list and the `mlx` extras group.

The two have to move together: mlx-vlm 0.7.0 and 0.7.1 require `mlx>=0.32.2`, so the raised ceiling admits nothing while mlx stays at 0.32.1. Under the existing transformers cap the default resolution is unchanged — 0.6.4 is still the newest mlx-vlm that fits, since 0.6.5 and later require `transformers>=5.14.0` — so the ceiling's effect is to let an installer that overrides transformers constrain this requirement rather than override it. mlx 0.32.1 and 0.32.2 publish the same set of macOS wheels.

## Risks and things a reviewer should weigh

- The decode-window guard hard-codes upstream's dispatch condition. That is deliberate and consistent with the existing contract: if upstream changes the condition, the body hash drifts and the fusion declines rather than producing wrong numbers.
- mlx-vlm now accepts two `SwitchGLU.__call__` bodies. Only the installed one can be verified at runtime; the pre-0.7.1 hash was verified against the v0.4.4 through v0.7.0 tags.
- The fused call delegates when `weights`, `shared` or `residual` arrive non-`None`. No 0.7.1 call site passes them today, so this costs nothing now and avoids a `TypeError` if one starts.

## Validation

```
pytest tests/test_mlx_moe_metal.py tests/test_mlx_gated_delta_vjp.py tests/test_mlx_decode_metal.py
114 passed, 3 skipped
```

Run against mlx 0.32.2, mlx-lm 0.31.3, mlx-vlm 0.7.1 on Apple Silicon Metal.

Beyond the suite, the upstream surfaces this package reaches into were checked at the source level against 0.7.1 and mlx 0.32.2:

- Every imported upstream symbol still resolves. Of 8024 methods present in both the 0.6.17 and 0.7.1 trees, 30 changed their parameter list; `SwitchGLU.__call__` is the only one reachable from here. The four imported `mlx_vlm.utils` and `prompt_utils` functions whose signatures moved only gained `**kwargs`.
- `mlx.nn.silu` and every mlx-lm switch layer fingerprint still match on mlx 0.32.2, so the decode convolution/SiLU fusion is unaffected; its contract still resolves against the real `Qwen3_5GatedDeltaNet` and its outputs are checked bitwise.
- mlx 0.32.2 adds a `force_fused` keyword to `mx.fast.scaled_dot_product_attention`. The fully-masked-row NaN workaround replaces that symbol, and its wrapper already forwards unknown keywords; verified that `force_fused` and `sinks` pass through bitwise unchanged and the masked-row guard still fires.
- mlx 0.32.2's new fused cross-entropy is gated on CUDA, so `nn.losses.cross_entropy` numerics are unchanged on Metal. Adam and Lion gained bounds checks that this package's defaults satisfy.
- The MiniCPM-o `sanitize` fingerprint drifts at 0.7.1, but 0.7.1 fixes the very defect that shim compensates for by accepting converted tower names, so the shim correctly goes inert and is left alone.

## Test fixture changes

The GLM-5 text config moves to an `index_topk` that 0.7.1's new `index_topk % index_kpool` check accepts, and the fused-input-projection test skips where that layer projects q, k and v through one weight — 0.7.1 removed the cached concatenation that test exercised, which leaves the production helper nothing to disable.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

File: unsloth_zoo/gated_delta_vjp.py [hunk 1]
```diff
@@ -454,12 +518,15 @@ def patch_gated_delta_vlm_shared():
     original_update = vlm_gated_delta.gated_delta_update
 
     def patched_shared_gated_delta_update(q, k, v, a, b, A_log, dt_bias,
-            state=None, mask=None, use_kernel=True, lower_bound=None):
-        if not _is_training_call(state, use_kernel):
-            # Pre-0.6.9 mlx-vlm has no `lower_bound` here, and no caller passes one.
-            bound = {} if lower_bound is None else {"lower_bound": lower_bound}
+            state=None, mask=None, use_kernel=True, **kwargs):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/gated_delta_vjp.py:521]: Checked every call site rather than the signature: nothing passes lower_bound positionally, in any supported version. In mlx-vlm 0.7.1 the five callers are models/gated_delta.py:367, glm5_next/language.py:202, kimi_k3/language.py:398 and :486, and qwen3_next/language.py:242, and each writes `lower_bound=...` as a keyword; kimi_linear/language.py:332 does not pass it at all. mlx-vlm 0.6.4 and mlx-lm 0.31.3 have no caller that passes it either. Positional through `state`, `mask` does happen and the wrapper still declares those three positionally, so those calls are unaffected.

Re-declaring lower_bound is also what the change is getting away from: the whole point of `**kwargs` here is that one wrapper stays correct from 0.4.x through 0.7.x without sniffing versions, and naming the keyword again reintroduces the coupling that 0.7.1's cache/cache_index/state_steps just broke.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
Real Apple Silicon differential against a base arm, since the repo's Metal job is opt-in behind the `mlx` label and the `latest` leg that installs 0.7.1 only runs on schedule or dispatch.

macos-15, the full `mlx-ci.yml` suite, three disposable branches - this PR's head, its merge base `b04baf02`, and base with only the mlx pin moved to 0.32.2 - each dispatched with `if: always()` on every test step, because the job runs each file as its own step and normally stops at the first red one.

| leg | base `b04baf02` | this PR |
|---|---|---|
| pinned (mlx-vlm 0.6.4) | `test_mlx_training_e2e_metal.py` fails | **entire suite green** |
| latest (mlx-vlm 0.7.1) | 28 failures: `test_qwen35_vjp_metal` 1, `test_mlx_moe_metal` 26, `test_mlx_generate_metal` 1 | **1 failure** |

The 26 MoE failures on base are the `SwitchGLU.__call__` fingerprint going stale on 0.7.1, which the per-package re-fingerprinting here clears. `test_end_to_end_training_step_all_patches` now passes on 0.7.1 too: it clears the `patched_vlm_gated_delta_update` assert and completes the training step, so the `cache=` `TypeError` is resolved end to end and not just at the signature.

The one remaining failure, `test_mlx_generate_metal.py::test_vlm_batched_generation_is_ordered_and_aligned`, fails identically on the base arm, so it is pre-existing under 0.7.1 and not this PR's doing.

The isolating arm settles the pinned-leg question: base code with only `mlx==0.32.1` changed to `0.32.2` is also fully green, so the bump alone fixes `test_mlx_training_e2e_metal.py` and none of the remaining movement is attributable to it.

One fix went in on top of the PR. Replacing the explicit `vlm_language.gated_delta_update = patched` with `_rebind_vlm_gated_delta_consumers` made the rebind identity-based, so it stopped reaching a consumer holding a different copy of the function than the one being replaced - which is exactly what `tests/test_qwen35_vjp_metal.py` produces, since it reloads `mlx_vlm.models.qwen3_5.gated_delta` before patching. That took `test_end_to_end_training_step_all_patches` red on both mlx-vlm versions. The sweep is kept, since it is what reaches `speculative_verifier` and `qwen4_exp`; the by-name bind is restored on top of it.

**Comment by danielhanchen** [comment]
Confirmed the shared SwitchGLU fingerprint in unsloth_zoo/mlx/inference.py:125 goes stale on mlx-vlm 0.7.1 so the gate/up packing stops applying, and that the fixed gated_delta_update signature raises on the new cache keywords; the Apple Silicon differential you ran matches what I see, will get this reviewed.
